### PR TITLE
Add Altos 586 emulation

### PR DESCRIPTION
**Pull request by lkundrak**
Hi. This adds incomplete emulation of Altos 586 -- the main board and a hard drive board. It is good enough to boot and pass tests on the diagnostic floppy, but not much else (no XENIX, CCP/M, OASIS). More details in the commit messages of the individual commits and code comments.

I eventually intend to extend the emulation to boot all the supported software, but as I'm pretty new to MAME, I thought it's a good idea to post this for review early, before I go too far doing something stupid. 

This is my first time programming C++ so please go easy on me :)

### CODE REVIEW

**Comment by lkundrak** [comment]
I realize I'm asking for your time and effort, but any chances I could get a review?
Anything I can do to help with that?
Thanks!

**Comment by rb6502** [comment]
@cracyc Can you look at the i86 change?  I'm gonna look at the rest, but I definitely want your eyes on that part :-)

**Comment by cracyc** [comment]
A glance at the microcode says that out byte and out word both put all of AX on the bus so seems okay.

**Comment by cuavas** [comment]
I’ll look at this.  It was marked draft, and GitHub doesn’t seem to send a notification when things are marked ready for review, so I didn’t see the change.

File: src/mame/altos/altos586_hdc.h [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+
+	uint16_t m_cyl_latch;
+	uint16_t m_cyl[2];
+
+	uint8_t m_sector[517];
+	int m_secoffset;
+
+	uint8_t m_drive;
+	uint8_t m_head;
+	const struct hard_disk_file::info* m_geom[2];
```
Comment (cuavas) [inline src/mame/altos/altos586_hdc.h:78]: Please keep the data members and member functions grouped in each section – you’ve got data members above and below the member functions here.

File: src/mame/altos/altos586_hdc.h [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+	void altos586_hdc_io(address_map &map);
+	void altos586_hdc_mem(address_map &map);
+
+private:
+	required_device<device_memory_interface> m_bus;
+	required_device<i8089_device> m_iop;
+	required_device<harddisk_image_device> m_hdd0;
+	required_device<harddisk_image_device> m_hdd1;
+
+	address_space *m_bus_mem;
```
Comment (cuavas) [inline src/mame/altos/altos586_hdc.h:47]: You should use `required_address_space` in place of `m_bus` and `m_bus_mem`.  It lets the user configure the address space number and takes care of validating that the address space exists.

File: src/mame/altos/altos586_hdc.cpp [hunk 3]
```diff
@@ -0,0 +1,380 @@
...
+	, m_hdd1(*this, "drive1")
+{
+}
+
+bool altos586_hdc_device::sector_exists(uint8_t index)
+{
+	if (!m_geom[m_drive]) {
+		logerror("drive not present: %d", m_drive);
+	} else if (m_geom[m_drive]->sectorbytes != 512) {
+		logerror("expected 512 bytes per sector, got %d", m_geom[m_drive]->sectorbytes);
```
Comment (cuavas) [inline src/mame/altos/altos586_hdc.cpp:74]: Should this be checked when mounting a disk image rather than on every sector access?

**Comment by cuavas** [comment]
This is structured pretty well and easy to read.  It should be easy enough to clean up and get it ready for merging.

**Comment by lkundrak** [comment]
I've also spent some time figuring out what is needed to make XENIX boot. With a fix to how the 8254 timer is wired up, it appears to sometimes (like 1 in 50 attempts) goes as far loading the kernel and having it print something!
Screenshot from 2024-03-20 17-25-37

File: src/mame/mame.lst [hunk 4]
```diff
@@ -581,6 +581,9 @@ altos486
 @source:altos/altos8600.cpp
 altos8600
 
+@source:altos/altos586.cpp
+altos586
+
 @source:amiga/alg.cpp
```
Comment (cuavas) [inline src/mame/mame.lst:587]: Please keep this file in order – this should come after the `altos/altos486.cpp` section.

File: src/mame/altos/altos586_hdc.cpp [hunk 5]
```diff
@@ -0,0 +1,371 @@
...
+
+uint16_t altos586_hdc_device::mem_r(offs_t offset, uint16_t mem_mask)
+{
+	return m_bus->read_word_unaligned(offset << 1, mem_mask);
+}
+
+void altos586_hdc_device::mem_w(offs_t offset, uint16_t data, uint16_t mem_mask)
+{
+	m_bus->write_word_unaligned(offset << 1, data, mem_mask);
+}
```
Comment (cuavas) [inline src/mame/altos/altos586_hdc.cpp:114]: Can you just use `read_word` and `write_word` here?  You’ll only get aligned memory accesses through the address space read/write handlers, so it shouldn’t be necessary to use the ones that have extra code for unaligned accesses.

File: src/mame/altos/altos586_hdc.cpp [hunk 6]
```diff
@@ -0,0 +1,371 @@
...
+	m_geom[index] = &m_hdd[index]->get_info();
+	return std::error_condition();
+}
+
+void altos586_hdc_device::device_add_mconfig(machine_config &config)
+{
+	I8089(config, m_iop, XTAL(15'000'000) / 3);
+	m_iop->set_addrmap(AS_PROGRAM, &altos586_hdc_device::altos586_hdc_mem);
+	m_iop->set_addrmap(AS_IO, &altos586_hdc_device::altos586_hdc_io);
+	m_iop->set_data_width(16);
```
Comment (cuavas) [inline src/mame/altos/altos586_hdc.cpp:340]: Does this board have its own clock, or does it use the bus clock for this?  If it uses the bus clock, you can use the `DERIVED_CLOCK` helper for the 8089 clock frequency and set the bus clock when instantiating the controller board.

File: src/mame/altos/altos586_hdc.h [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+
+	uint16_t m_cyl_latch;
+	uint16_t m_cyl[2];
+
+	uint8_t m_sector[517];
+	int m_secoffset;
+
+	uint8_t m_drive;
+	uint8_t m_head;
+	const struct hard_disk_file::info* m_geom[2];
```
Comment (cuavas) [inline src/mame/altos/altos586_hdc.h:83]: You don’t seem to be initialising `m_geom` in the constructor or `device_start`.

File: src/mame/altos/altos586_hdc.h [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+	// Disk access routines and state.
+	bool sector_exists(uint8_t index);
+	uint32_t sector_lba(uint8_t index);
+	void sector_read(uint8_t index);
+	void sector_write(uint8_t index);
+
+	// Image mount and unmount.
+	std::error_condition hdd_load(device_image_interface &image, uint8_t index);
+	std::error_condition hdd0_load(device_image_interface &image) { return hdd_load(image, 0); }
+	std::error_condition hdd1_load(device_image_interface &image) { return hdd_load(image, 1); }
```
Comment (cuavas) [inline src/mame/altos/altos586_hdc.h:63]: You can avoid the need for having three of these by making the drive index a template parameter.

File: src/mame/altos/altos586_hdc.cpp [hunk 9]
```diff
@@ -0,0 +1,371 @@
...
+
+void altos586_hdc_device::attn_w(uint16_t data)
+{
+	m_iop->ca_w(ASSERT_LINE);
+	m_iop->ca_w(CLEAR_LINE);
+}
+
+void altos586_hdc_device::device_start()
+{
+}
```
Comment (cuavas) [inline src/mame/altos/altos586_hdc.cpp:371]: You should at least do basic save state registration here (register you data members for save states).

File: src/mame/altos/altos586.cpp [hunk 10]
```diff
@@ -0,0 +1,775 @@
...
+	// Managed I/O access from the bus (IOP and HDC)
+	void bus_io(address_map &map);
+	u16 bus_io_r(offs_t offset, u16 mem_mask);
+	void bus_io_w(offs_t offset, u16 data, u16 mem_mask);
+
+	// device-level overrides
+	virtual void device_start() override;
+
+	// device_memory_interface overrides
+	virtual space_config_vector memory_space_config() const override;
```
Comment (cuavas) [inline src/mame/altos/altos586.cpp:117]: Overridden `protected` member functions should be `protected`.

File: src/mame/altos/altos586.cpp [hunk 11]
```diff
@@ -0,0 +1,775 @@
...
+	}
+
+	void altos586(machine_config &config);
+
+private:
+	// machine-level overrides
+	virtual void machine_start() override;
+
+	// device_memory_interface overrides
+	virtual space_config_vector memory_space_config() const override;
```
Comment (cuavas) [inline src/mame/altos/altos586.cpp:402]: Overridden `protected` member functions should be `protected`.

**Review by cuavas — COMMENTED** [review]
This looks pretty good.  Just a few things I noticed.

File: src/mame/altos/altos586_hdc.cpp (hunk 9 above)
Comment (lkundrak) [reply src/mame/altos/altos586_hdc.cpp]: Done, but I'm not sure it's correct -- I haven't yet used save/restore, so I need to figure out how to test it first.

**Pull request merged**